Flaky test: ServerSideXDS_InterceptorLeak_RDSUpdate

Logs:

<Details>

```
FAIL: Test (7.32s)
    --- FAIL: Test/ServerSideXDS_InterceptorLeak_RDSUpdate (0.01s)
        setup.go:45: Created new snapshot cache...
        tlogger.go:133: INFO server.go:732 [core] [Server #1918] Server created  (t=+272.382µs)
        setup.go:45: Registered Aggregated Discovery Service (ADS)...
        setup.go:45: xDS management server serving at: 127.0.0.1:44019...
        tlogger.go:133: INFO server.go:928 [core] [Server #1918 ListenSocket #1919] ListenSocket created  (t=+357.292µs)
        tlogger.go:133: INFO server.go:732 [core] [Server #1920] Server created  (t=+834.352µs)
        tlogger.go:133: INFO pool.go:309 [xds] xDS node ID: f37de6dd-8d6f-429f-89b5-0e6bee8f68a4  (t=+881.682µs)
        tlogger.go:133: INFO server.go:118 [xds] [xds-server 0xc0007aa380] Created xds.GRPCServer  (t=+899.068µs)
        stubserver.go:300: Started test service backend at "127.0.0.1:38775"
        tlogger.go:133: INFO server.go:177 [xds] [xds-server 0xc0007aa380] Serve() passed a net.Listener on 127.0.0.1:38775  (t=+1.049119ms)
        tlogger.go:133: INFO clientconn.go:1837 [core] original dial target is: "passthrough:///127.0.0.1:44019"  (t=+1.07698ms)
        tlogger.go:133: INFO clientconn.go:516 [core] [Channel #1921] Channel created for target "passthrough:///127.0.0.1:44019"  (t=+1.093805ms)
        tlogger.go:133: INFO clientconn.go:247 [core] [Channel #1921] parsed dial target is: resolver.Target{URL:url.URL{Scheme:"passthrough", Opaque:"", User:(*url.Userinfo)(nil), Host:"", Path:"/127.0.0.1:44019", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}}  (t=+1.112493ms)
        tlogger.go:133: INFO clientconn.go:248 [core] [Channel #1921] Channel authority set to "127.0.0.1:44019"  (t=+1.120945ms)
        tlogger.go:133: INFO server.go:928 [core] [Server #1920 ListenSocket #1922] ListenSocket created  (t=+1.166473ms)
        tlogger.go:133: INFO clientconn.go:620 [core] [Channel #1921] Channel Connectivity change to CONNECTING  (t=+1.234808ms)
        tlogger.go:133: INFO resolver_wrapper.go:211 [core] [Channel #1921] Resolver state updated: {
              "Addresses": [
                {
                  "Addr": "127.0.0.1:44019",
                  "ServerName": "",
                  "Attributes": null,
                  "BalancerAttributes": null,
                  "Metadata": null
                }
              ],
              "Endpoints": [
                {
                  "Addresses": [
                    {
                      "Addr": "127.0.0.1:44019",
                      "ServerName": "",
                      "Attributes": null,
                      "BalancerAttributes": null,
                      "Metadata": null
                    }
                  ],
                  "Attributes": null
                }
              ],
              "ServiceConfig": null,
              "Attributes": null
            } (resolver returned new addresses)  (t=+1.279825ms)
        tlogger.go:133: INFO balancer_wrapper.go:121 [core] [Channel #1921] Channel switches to new LB policy "pick_first"  (t=+1.297451ms)
        server.go:229: Created new resource snapshot...
        tlogger.go:133: INFO clientconn.go:923 [core] [Channel #1921 SubChannel #1923] Subchannel created  (t=+1.322688ms)
        logging.go:30: setting snapshot for node f37de6dd-8d6f-429f-89b5-0e6bee8f68a4
        server.go:235: Updated snapshot cache with resource snapshot...
        tlogger.go:133: INFO clientconn.go:421 [core] [Channel #1921] Channel exiting idle mode  (t=+1.34475ms)
        tlogger.go:133: INFO clientconn.go:1303 [core] [Channel #1921 SubChannel #1923] Subchannel Connectivity change to CONNECTING  (t=+1.364049ms)
        tlogger.go:133: INFO clientconn.go:1837 [core] original dial target is: "xds:///my-service"  (t=+1.371425ms)
        tlogger.go:133: INFO clientconn.go:516 [core] [Channel #1924] Channel created for target "xds:///my-service"  (t=+1.383223ms)
        tlogger.go:133: INFO clientconn.go:247 [core] [Channel #1924] parsed dial target is: resolver.Target{URL:url.URL{Scheme:"xds", Opaque:"", User:(*url.Userinfo)(nil), Host:"", Path:"/my-service", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}}  (t=+1.395ms)
        tlogger.go:133: INFO clientconn.go:248 [core] [Channel #1924] Channel authority set to "my-service"  (t=+1.406998ms)
        tlogger.go:133: INFO clientconn.go:1471 [core] [Channel #1921 SubChannel #1923] Subchannel picks a new address "127.0.0.1:44019" to connect  (t=+1.426336ms)
        tlogger.go:133: INFO clientconn.go:1303 [core] [Channel #1921 SubChannel #1923] Subchannel Connectivity change to READY  (t=+1.654177ms)
        tlogger.go:133: INFO clientconn.go:620 [core] [Channel #1921] Channel Connectivity change to READY  (t=+1.679855ms)
        logging.go:30: respond type.googleapis.com/envoy.config.listener.v3.Listener (requested [grpc/server?xds.resource.listening_address=127.0.0.1:38775]) version "" with version "1" and resources [grpc/server?xds.resource.listening_address=127.0.0.1:38775]
        logging.go:30: open watch 1 for type.googleapis.com/envoy.config.listener.v3.Listener map[grpc/server?xds.resource.listening_address=127.0.0.1:38775:{}] from nodeID "f37de6dd-8d6f-429f-89b5-0e6bee8f68a4", version "1"
        logging.go:30: respond type.googleapis.com/envoy.config.route.v3.RouteConfiguration (requested [routeName]) version "" with version "1" and resources [routeName]
        tlogger.go:133: INFO rds_handler.go:147 RDS watch for resource "routeName" received update: &xdsresource.RouteConfigUpdate{VirtualHosts:[]*xdsresource.VirtualHost{(*xdsresource.VirtualHost)(0xc0004dcc40)}, ClusterSpecifierPlugins:map[string]clusterspecifier.BalancerConfig{}, Raw:(*anypb.Any)(0xc00088c370)}  (t=+2.26887ms)
        logging.go:30: open watch 2 for type.googleapis.com/envoy.config.route.v3.RouteConfiguration map[routeName:{}] from nodeID "f37de6dd-8d6f-429f-89b5-0e6bee8f68a4", version "1"
        tlogger.go:133: INFO clientconn.go:620 [core] [Channel #1924] Channel Connectivity change to CONNECTING  (t=+2.30987ms)
        tlogger.go:133: INFO pool.go:309 [xds] xDS node ID: f37de6dd-8d6f-429f-89b5-0e6bee8f68a4  (t=+2.448415ms)
        tlogger.go:133: INFO xds_resolver.go:157 [xds] [xds-resolver 0xc0008e47e0] Creating resolver for target: xds:///my-service  (t=+2.46541ms)
        tlogger.go:133: INFO clientconn.go:1837 [core] original dial target is: "passthrough:///127.0.0.1:44019"  (t=+2.487162ms)
        tlogger.go:133: INFO clientconn.go:516 [core] [Channel #1927] Channel created for target "passthrough:///127.0.0.1:44019"  (t=+2.504648ms)
        tlogger.go:133: INFO clientconn.go:247 [core] [Channel #1927] parsed dial target is: resolver.Target{URL:url.URL{Scheme:"passthrough", Opaque:"", User:(*url.Userinfo)(nil), Host:"", Path:"/127.0.0.1:44019", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}}  (t=+2.523756ms)
        tlogger.go:133: INFO clientconn.go:248 [core] [Channel #1927] Channel authority set to "127.0.0.1:44019"  (t=+2.532219ms)
        tlogger.go:133: INFO clientconn.go:421 [core] [Channel #1924] Channel exiting idle mode  (t=+2.561171ms)
        tlogger.go:133: INFO clientconn.go:620 [core] [Channel #1927] Channel Connectivity change to CONNECTING  (t=+2.584516ms)
        tlogger.go:133: INFO resolver_wrapper.go:211 [core] [Channel #1927] Resolver state updated: {
              "Addresses": [
                {
                  "Addr": "127.0.0.1:44019",
                  "ServerName": "",
                  "Attributes": null,
                  "BalancerAttributes": null,
                  "Metadata": null
                }
              ],
              "Endpoints": [
                {
                  "Addresses": [
                    {
                      "Addr": "127.0.0.1:44019",
                      "ServerName": "",
                      "Attributes": null,
                      "BalancerAttributes": null,
                      "Metadata": null
                    }
                  ],
                  "Attributes": null
                }
              ],
              "ServiceConfig": null,
              "Attributes": null
            } (resolver returned new addresses)  (t=+2.640809ms)
        tlogger.go:133: INFO balancer_wrapper.go:121 [core] [Channel #1927] Channel switches to new LB policy "pick_first"  (t=+2.661199ms)
        tlogger.go:133: INFO clientconn.go:923 [core] [Channel #1927 SubChannel #1928] Subchannel created  (t=+2.685575ms)
        tlogger.go:133: INFO clientconn.go:421 [core] [Channel #1927] Channel exiting idle mode  (t=+2.70212ms)
        tlogger.go:133: INFO clientconn.go:1303 [core] [Channel #1927 SubChannel #1928] Subchannel Connectivity change to CONNECTING  (t=+2.724708ms)
        tlogger.go:133: INFO clientconn.go:1471 [core] [Channel #1927 SubChannel #1928] Subchannel picks a new address "127.0.0.1:44019" to connect  (t=+2.740241ms)
        tlogger.go:133: INFO clientconn.go:1303 [core] [Channel #1927 SubChannel #1928] Subchannel Connectivity change to READY  (t=+2.928474ms)
        tlogger.go:133: INFO clientconn.go:620 [core] [Channel #1927] Channel Connectivity change to READY  (t=+2.951909ms)
        logging.go:30: respond type.googleapis.com/envoy.config.listener.v3.Listener (requested [my-service]) version "" with version "1" and resources [my-service]
        logging.go:30: open watch 3 for type.googleapis.com/envoy.config.listener.v3.Listener map[my-service:{}] from nodeID "f37de6dd-8d6f-429f-89b5-0e6bee8f68a4", version "1"
        logging.go:30: respond type.googleapis.com/envoy.config.route.v3.RouteConfiguration (requested [route-my-service]) version "" with version "1" and resources [route-my-service]
        logging.go:30: open watch 4 for type.googleapis.com/envoy.config.route.v3.RouteConfiguration map[route-my-service:{}] from nodeID "f37de6dd-8d6f-429f-89b5-0e6bee8f68a4", version "1"
        logging.go:30: respond type.googleapis.com/envoy.config.cluster.v3.Cluster (requested [cluster-my-service]) version "" with version "1" and resources [cluster-my-service]
        logging.go:30: open watch 5 for type.googleapis.com/envoy.config.cluster.v3.Cluster map[cluster-my-service:{}] from nodeID "f37de6dd-8d6f-429f-89b5-0e6bee8f68a4", version "1"
        logging.go:30: respond type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment (requested [endpoints-my-service]) version "" with version "1" and resources [endpoints-my-service]
        logging.go:30: open watch 6 for type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment map[endpoints-my-service:{}] from nodeID "f37de6dd-8d6f-429f-89b5-0e6bee8f68a4", version "1"
        tlogger.go:133: INFO resolver_wrapper.go:211 [core] [Channel #1924] Resolver state updated: {
              "Addresses": null,
              "Endpoints": [],
              "ServiceConfig": {
                "Config": {
                  "Config": null,
                  "Methods": {}
                },
                "Err": null
              },
              "Attributes": {
                "\u003c%!p(xdsclient.clientKeyType=grpc.xds.internal.client.Client)\u003e": "\u003c0xc000a8a240\u003e",
                "\u003c%!p(xdsdepmgr.xdsClusterSubscriberKey={})\u003e": "\u003c0xc0003f03c0\u003e",
                "\u003c%!p(xdsresource.xdsConfigkey={})\u003e": "\u003c0xc0007e77c0\u003e",
                "\u003c%!p(resolver.csKeyType=grpc.internal.resolver.configSelector)\u003e": "\u003c0xc00025d880\u003e"
              }
            } (service config updated)  (t=+3.864346ms)
        tlogger.go:133: INFO balancer_wrapper.go:121 [core] [Channel #1924] Channel switches to new LB policy "xds_cluster_manager_experimental"  (t=+3.895171ms)
        tlogger.go:133: INFO clustermanager.go:58 [xds] [xds-cluster-manager-lb 0xc0007e7980] Created  (t=+3.911876ms)
        tlogger.go:133: INFO balancergroup.go:274 [xds] [xds-cluster-manager-lb 0xc0007e7980] Adding child policy of type "cds_experimental" for child "cluster:cluster-my-service"  (t=+3.922522ms)
        tlogger.go:133: INFO balancergroup.go:102 [xds] [xds-cluster-manager-lb 0xc0007e7980] Creating child policy of type "cds_experimental" for child "cluster:cluster-my-service"  (t=+3.931875ms)
        tlogger.go:133: INFO cdsbalancer.go:90 [xds] [cds-lb 0xc000878700] Created  (t=+3.940909ms)
        tlogger.go:133: INFO cdsbalancer.go:167 [xds] [cds-lb 0xc000878700] Received balancer config update: {
              "LoadBalancingConfig": null,
              "cluster": "cluster-my-service",
              "isDynamic": false
            }  (t=+3.951765ms)
        tlogger.go:133: INFO balancer.go:76 [xds] [priority-lb 0xc00025d960] Created  (t=+3.997092ms)
        tlogger.go:133: INFO balancergroup.go:528 [xds] [xds-cluster-manager-lb 0xc0007e7980] Balancer state update from child cluster:cluster-my-service, new state: {ConnectivityState:CONNECTING Picker:0xc0008d6870}  (t=+4.11098ms)
        tlogger.go:133: INFO balancerstateaggregator.go:147 [xds] [xds-cluster-manager-lb 0xc0007e7980] State update from sub-balancer "cluster:cluster-my-service": {ConnectivityState:CONNECTING Picker:0xc0008d6870}  (t=+4.122967ms)
        tlogger.go:133: INFO balancergroup.go:274 [xds] [priority-lb 0xc00025d960] Adding child policy of type "outlier_detection_experimental" for child "priority-0-0"  (t=+4.137669ms)
        tlogger.go:133: INFO balancergroup.go:102 [xds] [priority-lb 0xc00025d960] Creating child policy of type "outlier_detection_experimental" for child "priority-0-0"  (t=+4.14517ms)
        tlogger.go:133: INFO balancer.go:94 [xds] [outlier-detection-lb 0xc000633d10] Created  (t=+4.161494ms)
        tlogger.go:133: INFO clusterimpl.go:88 [xds] [xds-cluster-impl-lb 0xc00085ec80] Created  (t=+4.177648ms)
        tlogger.go:133: INFO clusterimpl.go:100 [xds] [xds-cluster-impl-lb 0xc00085ec80] xDS credentials in use: false  (t=+4.186852ms)
        tlogger.go:133: INFO weightedtarget.go:65 [xds] [weighted-target-lb 0xc0007e7ec0] Created  (t=+4.221773ms)
        tlogger.go:133: INFO balancer.go:88 [xds] [wrrlocality-lb 0xc0007ae540] Created  (t=+4.2667ms)
        tlogger.go:133: INFO balancergroup.go:274 [xds] [weighted-target-lb 0xc0007e7ec0] Adding child policy of type "round_robin" for child "{region=\"region-1\", zone=\"zone-1\", sub_zone=\"subzone-1\"}"  (t=+4.289604ms)
        tlogger.go:133: INFO balancergroup.go:102 [xds] [weighted-target-lb 0xc0007e7ec0] Creating child policy of type "round_robin" for child "{region=\"region-1\", zone=\"zone-1\", sub_zone=\"subzone-1\"}"  (t=+4.299789ms)
        tlogger.go:133: INFO roundrobin.go:56 [roundrobin] [0xc0007ae900] Created  (t=+4.310324ms)
        tlogger.go:133: INFO clientconn.go:923 [core] [Channel #1924 SubChannel #1931] Subchannel created  (t=+4.335011ms)
        tlogger.go:133: INFO balancergroup.go:528 [xds] [weighted-target-lb 0xc0007e7ec0] Balancer state update from child {region="region-1", zone="zone-1", sub_zone="subzone-1"}, new state: {ConnectivityState:CONNECTING Picker:0xc000a7a2c0}  (t=+4.366467ms)
        tlogger.go:133: INFO aggregator.go:253 [xds] [weighted-target-lb 0xc0007e7ec0] Child pickers with config: map[{region="region-1", zone="zone-1", sub_zone="subzone-1"}:weight:1,picker:0xc000a7a2c0,state:CONNECTING,stateToAggregate:CONNECTING]  (t=+4.385916ms)
        tlogger.go:133: INFO balancergroup.go:528 [xds] [priority-lb 0xc00025d960] Balancer state update from child priority-0-0, new state: {ConnectivityState:CONNECTING Picker:0xc000ae30f8}  (t=+4.403982ms)
        tlogger.go:133: INFO balancergroup.go:528 [xds] [xds-cluster-manager-lb 0xc0007e7980] Balancer state update from child cluster:cluster-my-service, new state: {ConnectivityState:CONNECTING Picker:0xc000ae30f8}  (t=+4.412615ms)
        tlogger.go:133: INFO balancerstateaggregator.go:147 [xds] [xds-cluster-manager-lb 0xc0007e7980] State update from sub-balancer "cluster:cluster-my-service": {ConnectivityState:CONNECTING Picker:0xc000ae30f8}  (t=+4.419565ms)
        tlogger.go:133: INFO balancerstateaggregator.go:191 [xds] [xds-cluster-manager-lb 0xc0007e7980] Child pickers: map[cluster:cluster-my-service:picker:0xc000ae30f8,state:CONNECTING,stateToAggregate:CONNECTING]  (t=+4.438153ms)
        tlogger.go:133: INFO clientconn.go:1303 [core] [Channel #1924 SubChannel #1931] Subchannel Connectivity change to CONNECTING  (t=+4.462169ms)
        tlogger.go:133: INFO clientconn.go:1471 [core] [Channel #1924 SubChannel #1931] Subchannel picks a new address "127.0.0.1:38775" to connect  (t=+4.484331ms)
        tlogger.go:133: INFO clientconn.go:1303 [core] [Channel #1924 SubChannel #1931] Subchannel Connectivity change to READY  (t=+4.692448ms)
        tlogger.go:133: INFO balancergroup.go:528 [xds] [weighted-target-lb 0xc0007e7ec0] Balancer state update from child {region="region-1", zone="zone-1", sub_zone="subzone-1"}, new state: {ConnectivityState:READY Picker:0xc0005e4680}  (t=+4.739728ms)
        tlogger.go:133: INFO aggregator.go:253 [xds] [weighted-target-lb 0xc0007e7ec0] Child pickers with config: map[{region="region-1", zone="zone-1", sub_zone="subzone-1"}:weight:1,picker:0xc0005e4680,state:READY,stateToAggregate:READY]  (t=+4.752797ms)
        tlogger.go:133: INFO balancergroup.go:528 [xds] [priority-lb 0xc00025d960] Balancer state update from child priority-0-0, new state: {ConnectivityState:READY Picker:0xc0005d8498}  (t=+4.767609ms)
        tlogger.go:133: INFO balancergroup.go:528 [xds] [xds-cluster-manager-lb 0xc0007e7980] Balancer state update from child cluster:cluster-my-service, new state: {ConnectivityState:READY Picker:0xc0005d8498}  (t=+4.78185ms)
        tlogger.go:133: INFO balancerstateaggregator.go:147 [xds] [xds-cluster-manager-lb 0xc0007e7980] State update from sub-balancer "cluster:cluster-my-service": {ConnectivityState:READY Picker:0xc0005d8498}  (t=+4.790333ms)
        tlogger.go:133: INFO balancerstateaggregator.go:191 [xds] [xds-cluster-manager-lb 0xc0007e7980] Child pickers: map[cluster:cluster-my-service:picker:0xc0005d8498,state:READY,stateToAggregate:READY]  (t=+4.798415ms)
        tlogger.go:133: INFO clientconn.go:620 [core] [Channel #1924] Channel Connectivity change to READY  (t=+4.811394ms)
        server.go:229: Created new resource snapshot...
        logging.go:30: setting snapshot for node f37de6dd-8d6f-429f-89b5-0e6bee8f68a4
        logging.go:30: respond open watch 5 type.googleapis.com/envoy.config.cluster.v3.Cluster [cluster-my-service] with new version "2"
        logging.go:30: respond type.googleapis.com/envoy.config.cluster.v3.Cluster (requested [cluster-my-service]) version "1" with version "2" and resources [cluster-my-service]
        logging.go:30: respond open watch 6 type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment [endpoints-my-service] with new version "2"
        logging.go:30: respond type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment (requested [endpoints-my-service]) version "1" with version "2" and resources [endpoints-my-service]
        logging.go:30: respond open watch 1 type.googleapis.com/envoy.config.listener.v3.Listener [grpc/server?xds.resource.listening_address=127.0.0.1:38775] with new version "2"
        logging.go:30: respond type.googleapis.com/envoy.config.listener.v3.Listener (requested [grpc/server?xds.resource.listening_address=127.0.0.1:38775]) version "1" with version "2" and resources [grpc/server?xds.resource.listening_address=127.0.0.1:38775]
        logging.go:30: respond open watch 2 type.googleapis.com/envoy.config.route.v3.RouteConfiguration [routeName] with new version "2"
        logging.go:30: respond type.googleapis.com/envoy.config.route.v3.RouteConfiguration (requested [routeName]) version "1" with version "2" and resources [routeName]
        logging.go:30: respond open watch 3 type.googleapis.com/envoy.config.listener.v3.Listener [my-service] with new version "2"
        logging.go:30: respond type.googleapis.com/envoy.config.listener.v3.Listener (requested [my-service]) version "1" with version "2" and resources [my-service]
        logging.go:30: respond open watch 4 type.googleapis.com/envoy.config.route.v3.RouteConfiguration [route-my-service] with new version "2"
        logging.go:30: respond type.googleapis.com/envoy.config.route.v3.RouteConfiguration (requested [route-my-service]) version "1" with version "2" and resources [route-my-service]
        server.go:235: Updated snapshot cache with resource snapshot...
        logging.go:30: open watch 7 for type.googleapis.com/envoy.config.cluster.v3.Cluster map[cluster-my-service:{}] from nodeID "f37de6dd-8d6f-429f-89b5-0e6bee8f68a4", version "2"
        tlogger.go:133: INFO rds_handler.go:147 RDS watch for resource "routeName" received update: &xdsresource.RouteConfigUpdate{VirtualHosts:[]*xdsresource.VirtualHost{(*xdsresource.VirtualHost)(0xc00080f600)}, ClusterSpecifierPlugins:map[string]clusterspecifier.BalancerConfig{}, Raw:(*anypb.Any)(0xc0006bc4b0)}  (t=+5.338973ms)
        xds_server_interceptor_leak_test.go:274: Destroyed 1 interceptor instances, want: 2
        tlogger.go:133: INFO clientconn.go:620 [core] [Channel #1924] Channel Connectivity change to SHUTDOWN  (t=+5.384756ms)
        tlogger.go:133: INFO resolver_wrapper.go:112 [core] [Channel #1924] Closing the name resolver  (t=+5.401781ms)
        tlogger.go:133: INFO balancer_wrapper.go:159 [core] [Channel #1924] ccBalancerWrapper: closing  (t=+5.417324ms)
        logging.go:30: open watch 8 for type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment map[endpoints-my-service:{}] from nodeID "f37de6dd-8d6f-429f-89b5-0e6bee8f68a4", version "2"
        logging.go:30: open watch 9 for type.googleapis.com/envoy.config.listener.v3.Listener map[my-service:{}] from nodeID "f37de6dd-8d6f-429f-89b5-0e6bee8f68a4", version "2"
        logging.go:30: open watch 10 for type.googleapis.com/envoy.config.route.v3.RouteConfiguration map[route-my-service:{}] from nodeID "f37de6dd-8d6f-429f-89b5-0e6bee8f68a4", version "2"
        logging.go:30: open watch 11 for type.googleapis.com/envoy.config.listener.v3.Listener map[grpc/server?xds.resource.listening_address=127.0.0.1:38775:{}] from nodeID "f37de6dd-8d6f-429f-89b5-0e6bee8f68a4", version "2"
        logging.go:30: open watch 12 for type.googleapis.com/envoy.config.route.v3.RouteConfiguration map[routeName:{}] from nodeID "f37de6dd-8d6f-429f-89b5-0e6bee8f68a4", version "2"
        tlogger.go:129: WARNING ads_stream.go:410 [xds] [xds-client 0xc0006de320] [xds-channel 0xc000784690] [ads-stream 0xc0006de3c0] Sending ADS request for type "type.googleapis.com/envoy.config.listener.v3.Listener", resources: [], version: "2", nonce: "7" failed: EOF  (t=+5.475179ms)
        tlogger.go:129: WARNING ads_stream.go:452 [xds] [xds-client 0xc0006de320] [xds-channel 0xc000784690] [ads-stream 0xc0006de3c0] ADS stream closed: rpc error: code = Canceled desc = context canceled  (t=+5.564767ms)
        tlogger.go:133: INFO ads_stream.go:165 [xds] [xds-client 0xc0006de320] [xds-channel 0xc000784690] [ads-stream 0xc0006de3c0] Shutdown ADS stream  (t=+5.575362ms)
        tlogger.go:133: INFO clientconn.go:620 [core] [Channel #1927] Channel Connectivity change to SHUTDOWN  (t=+5.586349ms)
        tlogger.go:133: INFO resolver_wrapper.go:112 [core] [Channel #1927] Closing the name resolver  (t=+5.594831ms)
        tlogger.go:133: INFO balancer_wrapper.go:159 [core] [Channel #1927] ccBalancerWrapper: closing  (t=+5.617174ms)
        tlogger.go:133: INFO clusterimpl.go:508 [xds] [xds-cluster-impl-lb 0xc00085ec80] Shutdown  (t=+5.630093ms)
        tlogger.go:133: INFO balancergroup.go:351 [xds] [priority-lb 0xc00025d960] Removing child policy for child "priority-0-0"  (t=+5.649752ms)
        tlogger.go:133: INFO cdsbalancer.go:416 [xds] [cds-lb 0xc000878700] Shutdown  (t=+5.658525ms)
        tlogger.go:133: INFO clustermanager.go:192 [xds] [xds-cluster-manager-lb 0xc0007e7980] Shutdown  (t=+5.666887ms)
        tlogger.go:133: INFO clientconn.go:1303 [core] [Channel #1927 SubChannel #1928] Subchannel Connectivity change to SHUTDOWN  (t=+5.674809ms)
        tlogger.go:133: INFO clientconn.go:1710 [core] [Channel #1927 SubChannel #1928] Subchannel deleted  (t=+5.684934ms)
        tlogger.go:133: INFO clientconn.go:1232 [core] [Channel #1927] Channel deleted  (t=+5.752754ms)
        tlogger.go:133: INFO channel.go:143 [xds] [xds-client 0xc0006de320] [xds-channel 0xc000784690] Shutdown  (t=+5.766324ms)
        tlogger.go:133: INFO xdsclient.go:210 [xds] [xds-client 0xc0006de320] Shutdown  (t=+5.780515ms)
        tlogger.go:133: INFO xds_resolver.go:295 [xds] [xds-resolver 0xc0008e47e0] Shutdown  (t=+5.789508ms)
        tlogger.go:133: INFO clientconn.go:1303 [core] [Channel #1924 SubChannel #1931] Subchannel Connectivity change to SHUTDOWN  (t=+5.820113ms)
        tlogger.go:133: INFO clientconn.go:1710 [core] [Channel #1924 SubChannel #1931] Subchannel deleted  (t=+5.828736ms)
        tlogger.go:133: INFO clientconn.go:1232 [core] [Channel #1924] Channel deleted  (t=+5.913281ms)
        tlogger.go:129: WARNING ads_stream.go:410 [xds] [xds-client 0xc0006a0960] [xds-channel 0xc00069e550] [ads-stream 0xc0006a0aa0] Sending ADS request for type "type.googleapis.com/envoy.config.listener.v3.Listener", resources: [], version: "2", nonce: "3" failed: EOF  (t=+5.976815ms)
        tlogger.go:129: WARNING ads_stream.go:452 [xds] [xds-client 0xc0006a0960] [xds-channel 0xc00069e550] [ads-stream 0xc0006a0aa0] ADS stream closed: rpc error: code = Canceled desc = context canceled  (t=+6.002813ms)
        tlogger.go:133: INFO ads_stream.go:165 [xds] [xds-client 0xc0006a0960] [xds-channel 0xc00069e550] [ads-stream 0xc0006a0aa0] Shutdown ADS stream  (t=+6.015743ms)
        tlogger.go:133: INFO clientconn.go:620 [core] [Channel #1921] Channel Connectivity change to SHUTDOWN  (t=+6.026438ms)
        tlogger.go:133: INFO resolver_wrapper.go:112 [core] [Channel #1921] Closing the name resolver  (t=+6.042172ms)
        tlogger.go:133: INFO balancer_wrapper.go:159 [core] [Channel #1921] ccBalancerWrapper: closing  (t=+6.062922ms)
        tlogger.go:133: INFO clientconn.go:1303 [core] [Channel #1921 SubChannel #1923] Subchannel Connectivity change to SHUTDOWN  (t=+6.091865ms)
        tlogger.go:133: INFO clientconn.go:1710 [core] [Channel #1921 SubChannel #1923] Subchannel deleted  (t=+6.099887ms)
        tlogger.go:133: INFO clientconn.go:1232 [core] [Channel #1921] Channel deleted  (t=+6.16975ms)
        tlogger.go:133: INFO channel.go:143 [xds] [xds-client 0xc0006a0960] [xds-channel 0xc00069e550] Shutdown  (t=+6.193835ms)
        tlogger.go:133: INFO server.go:864 [core] [Server #1920 ListenSocket #1922] ListenSocket deleted  (t=+6.212393ms)
        tlogger.go:133: INFO xdsclient.go:210 [xds] [xds-client 0xc0006a0960] Shutdown  (t=+6.238642ms)
        tlogger.go:133: INFO server.go:864 [core] [Server #1918 ListenSocket #1919] ListenSocket deleted  (t=+6.287263ms)
FAIL
```

</Details>

Failing run: https://github.com/grpc/grpc-go/actions/runs/26517271385/job/78097932567?pr=9118

--- a/test/xds/xds_server_interceptor_leak_test.go
+++ b/test/xds/xds_server_interceptor_leak_test.go
@@ -245,7 +245,11 @@ func (s) TestServerSideXDS_InterceptorLeak_RDSUpdate(t *testing.T) {
 
 	// Wait for the updated config to be applied. Since the new configuration has
 	// two routes, we expect two interceptor instances to be created for each
-	// filter chain, for a total of 6 interceptor instances.
+	// filter chain, for a total of 6 interceptor instances. We also wait for
+	// the old interceptors (one per filter chain, two total) to be destroyed,
+	// since destruction happens after the new interceptors are created within
+	// the same RDS update callback. Breaking only on interceptorsCreated==6
+	// would race with the subsequent interceptorsDestroyed increment.
 	for ; ctx.Err() == nil; <-time.After(defaultTestShortTimeout) {
 		if _, err := client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
 			t.Fatalf("EmptyCall() failed: %v", err)
@@ -254,7 +258,7 @@ func (s) TestServerSideXDS_InterceptorLeak_RDSUpdate(t *testing.T) {
 		case <-pathCh:
 		default:
 		}
-		if interceptorsCreated.Load() == 6 {
+		if interceptorsCreated.Load() == 6 && interceptorsDestroyed.Load() == 2 {
 			break
 		}
 	}

--- a/test/xds/xds_server_interceptor_leak_test.go
+++ b/test/xds/xds_server_interceptor_leak_test.go
@@ -271,12 +271,6 @@ func (s) TestServerSideXDS_InterceptorLeak_RDSUpdate(t *testing.T) {
 	if got, want := filtersCreated.Load(), int32(1); got != want {
 		t.Fatalf("Created %d filter instances, want: %d", got, want)
 	}
-	if got, want := interceptorsCreated.Load(), int32(6); got != want {
-		t.Fatalf("Created %d interceptor instances, want: %d", got, want)
-	}
-	if got, want := interceptorsDestroyed.Load(), int32(2); got != want {
-		t.Fatalf("Destroyed %d interceptor instances, want: %d", got, want)
-	}
 
 	stopServer()
 	if got, want := filtersDestroyed.Load(), int32(1); got != want {

